imap/imap.go

```diff
@@ -0,0 +1,353 @@
+		}()
+
+		var cmd *imap.Command
+		// find all the UIDs
+		cmd, err = findEmails(client, search, since)
+		if err != nil {
+			responses <- Response{Err: err}
+			return
+		}
+		// gotta fetch 'em all
+		getEmails(client, cmd, markAsRead, delete, responses)
+	}()
+
+	return responses, nil
+}
+
+func getEmails(client Client, cmd *imap.Command, markAsRead, delete bool, responses chan Response) {
+	seq := &imap.SeqSet{}
+	msgCount := 0
+	for _, rsp := range cmd.Data {
+		for _, uid := range rsp.SearchResults() {
+			msgCount++
+			seq.AddNum(uid)
+		}
+	}
+
+	// nothing to request?! why you even callin me, foolio?
+	if seq.Empty() {
+		return
+	}
+
+	fCmd, err := imap.Wait(client.UIDFetch(seq, "INTERNALDATE", "BODY[]", "UID", "RFC822.HEADER"))
+	if err != nil {
+		responses <- Response{Err: fmt.Errorf("unable to perform uid fetch: %s", err)}
+		return
+	}
+
+	var email Email
+	for _, msgData := range fCmd.Data {
+		msgFields := msgData.MessageInfo().Attrs
+
+		// make sure is a legit response before we attempt to parse it
+		// deal with unsolicited FETCH responses containing only flags
+		// I'm lookin' at YOU, Gmail!
+		// http://mailman13.u.washington.edu/pipermail/imap-protocol/2014-October/002355.html
+		// http://stackoverflow.com/questions/26262472/gmail-imap-is-sometimes-returning-bad-results-for-fetch
+		if _, ok := msgFields["RFC822.HEADER"]; !ok {
+			continue
+		}
+
+		email, err = NewEmail(msgFields)
+		if err != nil {
+			responses <- Response{Err: fmt.Errorf("unable to parse email: %s", err)}
+			return
+		}
+
+		responses <- Response{Email: email}
+
+		if !markAsRead {
+			err = removeSeen(client, imap.AsNumber(msgFields["UID"]))
+			if err != nil {
+				responses <- Response{Err: fmt.Errorf("unable to remove seen flag: %s", err)}
+				return
+			}
+		}
+
+		if delete {
+			err = deleteEmail(client, imap.AsNumber(msgFields["UID"]))
+			if err != nil {
+				responses <- Response{Err: fmt.Errorf("unable to delete email: %s", err)}
+				return
+			}
+		}
+	}
+	return
+}
+
+func deleteEmail(client Client, UID uint32) error {
+	return alterEmail(client, UID, "\\DELETED", true)
+}
+
+func removeSeen(client Client, UID uint32) error {
+	return alterEmail(client, UID, "\\SEEN", false)
+}
+
+func alterEmail(client Client, UID uint32, flag string, plus bool) error {
+	flg := "-FLAGS"
+	if plus {
+		flg = "+FLAGS"
+	}
+	fSeq := &imap.SeqSet{}
+	fSeq.AddNum(UID)
+	_, err := imap.Wait(client.UIDStore(fSeq, flg, flag))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// NewEmail will parse an imap.FieldMap into an Email. This
+// will expect the message to container the internaldate and the body with
+// all headers included.
+func NewEmail(msgFields imap.FieldMap) (Email, error) {
+
+	rawBody := imap.AsBytes(msgFields["BODY[]"])
+
+	rawBodyStream := bytes.NewReader(rawBody)
+	em, err := email.NewEmailFromReader(rawBodyStream) // Parse with @jordanwright's library
+	if err != nil {
+		log.Error("Unable to parse email")
```

Comment:
It seems like we'd probably want to return the error here, right? I could be wrong, so I'd like to get your take on it.

Reply:
Right you are!